Allow building the cross compiler with MSVC

ARF1 opened the pull request:
This PR adds a build configuration to allow building mpy-cross using MSVC.

stinos commented on mpy-cross/README.md:
```diff
@@ -2,15 +2,28 @@ MicroPython cross compiler
 
     $ ./mpy-cross foo.py
+    
+On windows invoke it as:
```
This is basically the same as for the unix-like build so doesn't really need to be repeated imo

stinos commented:
Just duplicating most of the build files and adjusting a couple of lines here and there feels wrong. It's not interesting to maintain a couple of hundreds of lines of two nearly identical versions of the same set of files. I'd suggest to instead add the bare minimum needed (the project file, possible a .props file with customizations if needed) and if needed adjust the current msvc build files so they can be used to build other executables as well.

stinos commented:
@ARF1 any plans with this? Otherwise I suggest it gets closed as it's currently not a good fit for merging.

ARF1 commented:
@stinos I tried to implement the suggestion from your previous comment but I cannot get that to work. It is a minor miracle for me that I managed to get it to work as is. I am really out of my depth here.

It seems a shame though to throw away something that works without a replacement. Especially since installing the gcc build environment on windows is far from painfree!

How about leaving this open in the hope that somebody comes along that actually knows what he is doing with MSVC and fixes it? Once it is closes it effectively disappears...

An even better alternative would be if we closed this but compiled Windows cross compilers were offered as downloads on [the micropython download page](http://micropython.org/download).

stinos commented:
> It seems a shame though to throw away something that works without a replacement.

Maybe I'm doing it wrong but it doesn't work for me. I rebased this branch on master, built micropython and mpy-cross exes, made a file foo.py with one single function like `def X(): pass`, ran 'mpy-cross foo.py -o foo2.mpy', then ran micropython itself:
```
>>> import foo2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: incompatible .mpy file
```

> How about leaving this open in the hope that somebody comes along that actually knows what he is doing with MSVC and fixes it?

It is somwehere on my todo list but not at the top and I'm sort of drowning in work and hobbies at the moment so can take a while. And adding extra debugging to see why it doesn't seem to work as expected just adds to that :]

ARF1 commented:
@stinos 

> import foo2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: incompatible .mpy file

I remember I had a similar problem when the code base of the cross-compiler and the code base of the firmware installed on my device were different. I fixed it my rebasing the cross-compiler PR on the tag of the firmware. - But I am sure you already have compatible revisions.

I will try to find the time to take a look whether it still works for me.

stinos committed:
windows/msvc: Move some build options from the .vcxproj to .props files

We want the .vcxproj to be just a container with the bare minimum
content for making it work as a project file in Visual Studio, whereas
the actual build options and code gets placed in seperate small and
reusable files. This was roughly the case already except few compiler
options were overlooked; fix this here: we'll need those common
options when adding a project file for building mpy-cross.

stinos committed:
py: Allow windows ports to save persistent code

stinos committed:
windows: Default to binary mode for files

If this is not set it might default to calls to open() to use text
mode which is usually not wanted, and even completely wrong and leading
to incorrect results when loading the binary .mpy files.

stinos committed:
windows/msvc: Allow modification of include directories

Instead of just overwriting the include directories used, append
to them so project files including this file can easily define
include directories for the compiler.

stinos committed:
mpy-cross: Make mpconfigport.h compatible with msvc compiler

This is just a copyof the required options from windows/mpconfigport.h

stinos commented:
@ARF1 I made some progress for this, can you check if https://github.com/stinos/micropython/tree/mpy-cross-msvc works for you?

Note that to use the .mpy files in unix or windows ports, MICROPY_OPT_CACHE_MAP_LOOKUP_IN_BYTECODE must be set to 0 in mpconfigport.h: if not you get the `incompatible .mpy file` error.

dpgeorge commented:
> Note that to use the .mpy files in unix or windows ports, MICROPY_OPT_CACHE_MAP_LOOKUP_IN_BYTECODE must be set to 0 in mpconfigport.h: if not you get the incompatible .mpy file error.

There is an option to mpy-cross to make the compiled bytecode support this option:
```
$ mpy-cross -mcache-lookup-bc myfile.py
```
The single mpy-cross executable should work for all ports, regardless of their mpconfigport.h settings.  You just need to pass the right options to mpy-cross :)

stinos commented:
Oops, sorry, completely missed that!

stinos committed:
windows/msvc: Change the way sources are listed

Reserve sources.props for just the micropython 'core' files, similar
like how py/py.mk lists port-independent source files.
This allows reusing them, for instance for building mpy-cross.
The sources for building the micropython executable are listed in
the project file itself, similar to how the other ports specify their
source files in their Makefile.

stinos committed:
windows: Add msvc project file for building mpy-cross

This project buils the mpy-cross application for themsvc port.
Note the build/output directories are under windows/ directory
instead of mpy-cross directory (and hence added to windows/.gitignore)

Commit:
windows/msvc: Do not attempt to compile machine_pulse.c when building mpy-cross

Moved extmod\machine_pulse.c from sources.props to micropython.vcxproj to only
build with micropython. When building mpy-cross the file fails to compile.

ARF1 commented:
@stinos I missed your message from last year on this. Thank you very much for taking the time to refactor it into a sensible form! MSVC build configs look a bit like chinese to me; I would not have been able to get there on my own.

I took your improved code and rebased it to restore mergeability. (I edited your changesets during the rebase to fit with the new directory structure.)

Note: I had to remove the reference to `extmod\machine_pulse.c` from `sources.prop` and move it back to `micropython.vcxproj`. This file compiles perfectly fine when building `micropython` on MSVC but refuses to compile when building `mpy-cross`:
```
ClCompile:
  C:\Program Files\Microsoft Visual Studio 14.0\VC\bin\cl.exe /c /I"Z:\Programming\esp8266\micropython\ports\windows\\..\..\mpy-cross" /IZ:\Programming\esp8266\micropython\ /IZ:\Programming\esp8266\micropython\ports\windows\ /I"Z:\Programming\esp8266\micropython\ports\windows\\buildmpy-cross\\" /IZ:\Programming\esp8266\micropython\ports\windows\msvc /ZI /nologo /W1 /WX- /sdl- /MP /Od /Oy- /D _USE_MATH_DEFINES /D _CRT_SECURE_NO_WARNINGS /D _MBCS /Gm- /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"Z:\Programming\esp8266\micropython\ports\windows\\buildmpy-cross\DebugWin32\obj\\" /Fd"Z:\Programming\esp8266\micropython\ports\windows\\buildmpy-cross\DebugWin32\obj\vc140.pdb" /Gd /TC /analyze- /errorReport:queue Z:\Programming\esp8266\micropython\extmod\machine_pulse.c machine_pulse.c
z:\programming\esp8266\micropython\extmod\machine_pulse.h(32): error C2122: 'pulse_level': prototype parameter in name list illegal [Z:\Programming\esp8266\micropython\ports\windows\mpy-cross.vcxproj]
z:\programming\esp8266\micropython\extmod\machine_pulse.h(32): error C2122: 'timeout_us': prototype parameter in name list illegal [Z:\Programming\esp8266\micropython\ports\windows\mpy-cross.vcxproj]

```
Since `extmod\machine_pulse.c` is not needed I took the easy way out and simply removed it from the build recipe for `mpy-cross`.

When you have spare time, would you take a look if my adaptation of your work is ok? From my end this PR looks ready.

stinos commented:
I'll check it out shortly. One thing I noticed already: there are missing spaces in at leat 2 commit messages ('copyof' 'themsvc')

stinos commented:
@ARF1 I reviewed this completely since it has been a while, updated changes in my mpy-cross-msvc branch:
- reworded some commits and make them adhere to the standard format with a dot at the end of the subject line
- copied all msvc options from windows/mpconfigport.h to mpy-cross/mpconfigport.h so it's easier in the future to update
- changed sources.props to list core and extmod files seperately, which automatically gets rid of the error in machine_pulse.c you got because that file is not included in the mpy-cross sources anymore
- add fmode.c to windows/Makefile else build fails
- fix appveyor build, and also make it build mpy-cross using mingw for the sake of continuously checking if that works; does require hacking around #3310 but that is non-trivial to fix so in the meantime I think it's still better to cover the mingw mpy-cross then to leave any possible errors undetected. Maybe this should go in a seperate commit though to make it clear what has changed.

If there are no objections to the above changes I think this is almost good to go, except for one thing to discuss: currently mpy-cross.exe ends up in the ports/windows directory which is inconsistent and possibly confusing since it normally ends up in the mpy-cross directory. Best to fix that as well then?

ARF1 commented:
@stinos I looked at your modifications but in all honesty I do not have the knowledge to comment on the specifics. I can tell however, that it still builds on my machine. :-)

Regarding the location of `mpy-cross.exe`: In my opinion, `mpy-cross.exe` should be created in the directory in which `mpy-cross.vcxproj` resides. Least confusing for users would be if both were in the `mpy-cross` directory because that is where a new user will look first if he want to build the cross compiler (whatever os he is on). But in the alternative the current solution of having both in the windows port directory is also quite reasonable. 

If `mpy-cross.exe` ends up in another directory than that from which the build was initiated, a new user will be very confused if he does not find the result of a successful build.

Improvements for novice users (in case you want to take this opportunity to improve accessibility):

1. If you keep `mpy-cross.vcxproj` in the windows port directory, I might be worth considering placing a short text file `How to build on windows.txt` in the `mpy-cross` directory to point the user to the project file in the windows port directory.

2. A hint on how to build might be useful in any case. When I first started, I was very confused that Visual Studio GUI would not open the `vcxproj` files at all. (On my system opening the project fails with the error: `The system cannot find the file specified.`) It took me a good while to figure out, that it was still possible to build them on the command line...

Are you able to update this PR yourself or shall I keep pushing your changesets to this PR? Please let me know if there is anything I can do to help.

stinos commented:
> Least confusing for users would be if both were in the mpy-cross directory because that is where a new user will look first if he want to build the cross compiler

Sounds reasonable indeed. And solves the issue of the mpy-cross.exe file location as well.
@dpgeorge any objections to placing another build file next to the makefile in mpy-cross?

> I was very confused that Visual Studio GUI would not open the vcxproj files at all

Probably worth creating a seperate issue for this including needed info (which version of VS, how exactly do you open the file, ...). We've been using the project files on different machines with 3 or 4 different versions of VS and never got any problems.

> Are you able to update this PR yourself

Probably not, but it's going to be easier if I just create a PR from my branch.

stinos commented:
@dpgeorge ping; IIRC all of this was pretty much ready, only decision left to make is the location of mpy-cross.exe and the project file, see discussion above.

dpgeorge commented:
> only decision left to make is the location of mpy-cross.exe and the project file,

If it's just a single `mpy-cross.vcxproj` file that would go in `mpy-cross/`, then I'm +1 for that option.  Makes sense to be able to build mpy-cross under MSVC separately.

dpgeorge commented on mpy-cross/mpconfigport.h:
```diff
@@ -124,3 +127,33 @@ typedef long mp_off_t;
+#ifdef _MSC_VER
+
+#define NORETURN                    __declspec(noreturn)
+#undef MP_NOINLINE
```
Should be able to remove this undef now due to recent c066dadc5b91d000e2f7bba9f3e3edbf0bf33a3d

stinos replied:
Confirmed

dpgeorge commented on ports/windows/init.c:
```diff
@@ -61,6 +62,7 @@ void init() {
     // https://msdn.microsoft.com/en-us/library/bb531344(v=vs.140).aspx
     _set_output_format(_TWO_DIGIT_EXPONENT);
 #endif
+    set_fmode_binary();
```
Is this needed anymore?  In 9bdb82ef6b3147662ffc6b6435c6bd376c60e19e a call to this was added in mpy-cross itself.  And wouldn't the addition here change the behaviour of the windows port of MicroPython?

stinos replied:
Without this loading persistent code might fail because the binary .mpy file gets read while having CRLF replaced by LF etc; assuming the default mode for `open()` is text which does seem to be the case. So that indeed changes behaviour, as also briefly discussed in #2209, but I think this change is actually good: problem now is that writing `\n` to a file in uPy yields `\r\n` - which likely went unnoticed so far just because reading is symmetrical and does the opposite. I'll address this in the commit message.

mpy-cross doesn't call `init()` currently so is unaffected by this change, but it's probably better to change mpy-cross so it does the same initialization.

dpgeorge commented on py/persistentcode.c:
```diff
@@ -377,7 +377,7 @@ void mp_raw_code_save(mp_raw_code_t *rc, mp_print_t *print) {
 // TODO abstract this away properly
 
-#if defined(__i386__) || defined(__x86_64__) || defined(__unix__)
+#if defined(__i386__) || defined(__x86_64__) || defined(_WIN32) || defined(__unix__)
```
Does this need a `_WIN64` as well?

stinos replied:
No, despite the `32` there, _WIN32 is defined as 1 when the compilation target is 32-bit ARM, 64-bit ARM, x86, or x64

stinos commented:
With the addition of native emitting mpy-cross doesn't compile anymore when `MICROPY_GCREGS_SETJMP` is 1 because in that case `nlr_push` is a macro and as such cannot be put in `mp_fun_table`. This doesn't seem trivial to fix though, i.e. it doesn't look like native emitting can work without it? Which would mean that to get mpy-cross built for the windows port, native emitting would have to be excluded from it. Is that problematic?

dpgeorge commented:
> With the addition of native emitting mpy-cross doesn't compile anymore when `MICROPY_GCREGS_SETJMP` is 1 because in that case `nlr_push` is a macro and as such cannot be put in `mp_fun_table`. This doesn't seem trivial to fix though

Native code doesn't actually need to run as part of mpy-cross so it should be possible to just exclude `mp_fun_table` outright and it'll still work the same.  All it needs is a few more #if's ... I'll look into it.

dpgeorge commented:
> Native code doesn't actually need to run as part of mpy-cross so it should be possible to just exclude `mp_fun_table` outright and it'll still work the same.

@stinos see a4f1d82757b8e95c21a095c99b7c3f04ded88104

stinos commented:
Ok thanks, that was easy :)

dpgeorge commented:
Superseded by #4826
